adr/0026-schema.md

```diff
@@ -0,0 +1,578 @@
+- *when* the user runs `zarf package create`
+- *then* they will receive an error and be told to run `zarf dev update-schema` or how to migrate off cosign key paths or how to use flavors over groups depending on the error
+
+#### v0 create -> v1 deploy
+- *Given*: A package is created with Zarf v0
+- *and* that package has deprecated keys that can be automatically migrated (required, scripts, & set variables)
+- *when* the package is deployed with Zarf v1
+- *then* the keys will be automatically migrated & the package will be deployed without error.
+
+#### v0 create with removed feature -> v1 deploy
+- *Given*: A package is created with Zarf v0
+- *and* that package has deprecated keys that cannot be automatically migrated (groups, cosignKeyPath)
+- *when* the package is deployed with Zarf v1
+- *then* then deploy of that package will fail and the user will be instructed to update their package
+
+#### v1 create -> v0 deploy
+- *Given*: A package is created with Zarf v1
+- *and* that package uses keys that don't exist in v0
+- *when* the package is deployed with Zarf v0
+- *then* Zarf v0 will deploy the package without issues. If there are fields unrecognized by the v0 schema, then the user will be warned they are deploying a package that has features that do not exist in the current version of Zarf.
+
+## Consequences
+- As long as the only deprecated features in a package have migration path, and the package was built after the feature was deprecated so migrations were run, Zarf will be successful both creating a package with v1 and deploying with prev1, and creating a package with prev1 and deploying with v1.
+- Users of deprecated group or cosignKeyPath might be frustrated if their packages, created prev1, error out on Zarf v1, however this is likely preferable to unexpected behavior occurring in the cluster.
+- Users may be frustrated that they have to run `zarf dev update-schema` to edit their `zarf.yaml` to remove the deprecated fields and add `apiVersion`.
+- We will have to have two different schema types which will be mostly be duplicate code. However the original type should never change, which mitigates much of the issue.
+- By having a zarf.yaml and a zarfv1.yaml it will be easy to read and write from objects to yamls directly without having to include deprecated v0 fields in the v1 schema. However, this will also mean any new keys in v1 won't exist in the `zarf.yaml` so a v0 deploy of a v1 package will not be able to warn users of unrecognized keys, we will have to use some other method.
+
+
+Below is an example v1 zarf.yaml with, somewhat, reasonable & nonempty values for every key
+```yaml
+kind: ZarfPackageConfig
+apiVersion: v1
+metadata:
+  name: everything-zarf-package
+  description: A zarf package with a non empty value for every
+  version: v1.0.0
+  url: https://my-package-website.com
+  image: https://my-image-url-to-use-in-deprecated-zarf-ui # TODO This field should be deprecated
```

Comment:
Gotcha the json description mentioned it was for the Zarf UI which is why I added that note, but that's good to know. We will keep it in.

Reply:
@jeff-mccoy Can you give me details or an example of what it'll be used for? I want to update the json description to be more clear

Reply:
Is there an information on how this field is helpful to Zarf + UDS Marketplace, not just UDS side of things?

Reply:
Yeah I don't think it's just marketplace or zarf ui. It's helpful data to package in airgap that someone could use to show info about the package in a common way. Having an image for a package is very common for any tool listing application in a catalog, it could be a logo or screenshot. 

Leaving that in a standard place and having zarf load that on package create is something I still think we should do for the ux for anyone who builds any kind of package catalog.

Reply:
One thought that was brought up would be a labels or annotations field. This would cover most of these use cases where users want to add arbitrary metadata. This is a common practice in other projects and would allow for more features than just images. Does that sound like a good idea @jeff-mccoy?

Reply:
One small note if image stays as is for the airgap use case - it would need to be baked into the package somehow for it to show to a user there.